library/psa_crypto.c

```diff
@@ -3944,17 +3961,21 @@ static psa_status_t psa_tls12_prf_set_seed( psa_tls12_prf_key_derivation_t *prf,
 }
 
 static psa_status_t psa_tls12_prf_set_key( psa_tls12_prf_key_derivation_t *prf,
-                                           psa_algorithm_t hash_alg,
                                            const uint8_t *data,
                                            size_t data_length )
 {
-    psa_status_t status;
     if( prf->state != PSA_TLS12_PRF_STATE_SEED_SET )
         return( PSA_ERROR_BAD_STATE );
 
-    status = psa_hmac_setup_internal( &prf->hmac, data, data_length, hash_alg );
-    if( status != PSA_SUCCESS )
-        return( status );
+    if( data_length != 0 )
+    {
+        prf->secret = mbedtls_calloc( 1, data_length );
```

Comment:
I'd say that this dynamic allocation is a problem as we want to avoid them as much as possible and there was no dynamic allocation with the previous code. Do we have a maximum for `data_length` to allocate a static array in the context instead ? Would it rather make sense to introduce an `mbedtls_psa_mac_clone` or `mbedtls_psa_hmac_clone` with associated driver interfaces ? @gilles-peskine-arm what do you think about that ?

Reply:
There was most certainly dynamic allocation with the previous code as well, in order to store seed and label. Ref. the respective setter functions.

The only way it would make sense to require non-dynamic-allocation is if the whole PRF operation can be done without dynamic allocation. Unless the plan is to allocate static buffers (and thus imposing size restrictions) for the seed and label bytestrings as well, I'd say there's no point in bending over backwards to avoid allocating the key buffer, since the effort would be inconsequential.

Reply:
ok I don't know well the PRF code. If there are already dynamic allocations here then that is probably fine. Gilles knows probably this part better than me. Let's wait and see what he thinks about it.

Reply:
Extending this requirements to operations which reference keys might be a mistake, since for opaque keys a driver might be doing its own reference counting.